File: lab1/tcprecv.py
#!/usr/bin/env python3
# tcprecv.py - TCP server with host and port arguments

import socket
import sys
import argparse

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='TCP server')
    parser.add_argument('--host', default='127.0.0.1', help='Host address to bind to')
    parser.add_argument('--port', type=int, default=8888, help='Port to listen on')
    args = parser.parse_args()

    # Create TCP socket
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)  # Allow reuse
    except socket.error as e:
        print(f"Socket creation error: {e}")
        sys.exit(1)

    # Bind socket
    try:
        s.bind((args.host, args.port))
        print(f"Bound to {args.host}:{args.port}")
    except socket.error as e:
        print(f"Bind error: {e}")
        s.close()
        sys.exit(1)

    # Listen for connections
    s.listen(5)  # Queue up to 5 connections
    print(f"Server listening on {args.host}:{args.port}...")

    try:
        while True:
            print("Waiting for incoming connections...")
            
            # Accept connection
            try:
                ns, client_addr = s.accept()
                print(f"Connection accepted from {client_addr[0]}:{client_addr[1]}")
            except socket.error as e:
                print(f"Accept failed: {e}")
                continue

            # Handle client connection
            try:
                while True:
                    # Receive data
                    data = ns.recv(512)  # BUFLEN = 512
                    if not data:
                        print("Client disconnected")
                        break
                    
                    # Remove null terminator if present
                    if data.endswith(b'\0'):
                        message = data[:-1].decode('utf-8', errors='replace')
                    else:
                        message = data.decode('utf-8', errors='replace')
                    
                    print(f"Got message {len(data)} bytes: {message}")
                    
                    # Check for session end
                    if message.startswith('.'):
                        print("Session ended by client")
                        break
                        
            except socket.error as e:
                print(f"Communication error: {e}")
            finally:
                ns.close()
                print("Client connection closed")
                
    except KeyboardInterrupt:
        print("\nServer shutting down...")
    finally:
        s.close()
        print("Server closed")

if __name__ == "__main__":
    main()

File: lab1/tcprecv.c
// tcprecv.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#define BUFLEN 512      // Max length of buffer
#define PORT 8888       // The port on which to send or listen for data
#define LOCAL_LOOP "127.0.0.1" 

void die(char *s)      // exit with message
{
        perror(s); exit(1);
}

int main(void)
{
  struct sockaddr_in si_me, si_other;
  int s,ns, nb, slen = sizeof(si_other);
  char buf[BUFLEN];
  s=socket(AF_INET, SOCK_STREAM, 0);
  memset((char *) &si_other, 0, sizeof(si_other));
  si_me.sin_family = AF_INET;
  si_me.sin_port = htons(PORT);
  //si_me.sin_addr.s_addr = htonl(INADDR_ANY);
  si_me.sin_addr.s_addr = inet_addr(LOCAL_LOOP);
  // Bind is mandatory for the server
  if( bind(s,(struct sockaddr *)&si_me , sizeof(si_me)) < 0) die("bind");
  listen(s,5);   // log length
  while(1)
    {
    //Accept and incoming connection
    puts("Waiting for incoming connections...");
    ns=accept(s, (struct sockaddr *)&si_other,(socklen_t*)&slen);
    if (ns<0) { puts("accept failed"); continue; }
    else  puts("Connection accepted");
    while(1)
      {
      memset(buf,0x00,BUFLEN);
      nb=read(ns,buf,BUFLEN);
      printf("Got message %d bytes: %s\n",nb,buf); if(nb==0) break;
      if(buf[0]=='.') { printf("session ended\n");break;}
      }
    close(ns);  // close the working socket
    }
  close(s);
}

